Replace assertions with ValueErros on distilbert model

JuheonChu opened the pull request:
Co-author: @batese2001
This is an extended PR from [here](https://github.com/huggingface/transformers/pull/20432)  for new valid-checks to improve _code quality._

This raises exceptions on Multi-head Attention models for following specified conditions on the _Distilbert_ model.

This successfully replaces assertions with ValueErrors with the _Distilbert_ model related to #12789.

Would you be open to me changing assertions if I encounter other ones?

To: @younesbelkada

Commit:
Changed assert into 7-8 exceptions

Commit:
updated syntax error

Commit:
updated error

JuheonChu committed:
Successful test on test_modeling_distilbert.py 

Successful raising errors and exceptions on the revised code in test_modeling_distilbert.py .

Co-credit: @batese2001

JuheonChu committed:
Delete test_modeling_distilbert.ipynb

JuheonChu committed:
Update modeling_distilbert.py

Commit:
updated file (Co-autho: Batese2001)

Commit:
Successful raising of exceptions with the conditions that are contrary to defined condition that asserts statements (Co-author: Batese2001)

Commit:
Successful raising of exceptions with the conditions that are contrary to defined condition that asserts statements (Co-author: Batese2001)

Commit:
committing the reformatted distilbert model

Commit:
reformatted distilbert model

Commit:
reformatted distilbert model

Commit:
reformatted distilbert model

Commit:
reformatted distilbert model with black

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20433). All of your documentation changes will be reflected on that endpoint.

JuheonChu commented:
@[Sylvain Gugger](https://github.com/sgugger) 
This is a new PR of my own based on your suggestion from [here](https://github.com/huggingface/transformers/pull/20375). files which improved to pass all the validity checks. 
Can I ask for your further suggestions for this PR to be merged? 
Any comments will be very appreciated!

Commit:
Changed comments that explain better about raising exceptions for not having the even number of multi heads

younesbelkada commented:
Hi @JuheonChu ! 👋
Thanks for the PR ;) will review it asap and let you know!

younesbelkada commented on src/transformers/models/distilbert/modeling_distilbert.py:
```diff
@@ -141,7 +142,11 @@ def __init__(self, config: PretrainedConfig):
+        # Have an even number of multi heads that divide the dimensions
+        if self.dim % self.n_heads != 0:
+            raise ValueError(
+                "self.n_heads (" + str(self.n_heads) + ") must divide self.dim (" + str(self.dim) + ") evenly"
```
```suggestion
                f"self.n_heads: {self.n_heads} must divide self.dim: {str(self.dim)} evenly"
```
It is preferable to use f-strings ;) check for example [here](https://github.com/huggingface/transformers/blob/3c39c07f1160beb6190f635e2dab6a27e6d67596/src/transformers/models/opt/modeling_opt.py#L139)

younesbelkada commented on src/transformers/models/distilbert/modeling_distilbert.py:
```diff
@@ -255,7 +260,11 @@ class TransformerBlock(nn.Module):
+        # Have an even number of Configure multi-heads
+        if config.dim % config.n_heads != 0:
+            raise ValueError(
+                "config.n_heads (" + str(config.n_heads) + ") must divide config.dim (" + str(config.dim) + ") evenly"
```
```suggestion
                f"config.n_heads {config.n_heads} must divide config.dim {config.dim} evenly"
```
same as above!

younesbelkada commented on src/transformers/models/distilbert/modeling_distilbert.py:
```diff
@@ -291,7 +300,9 @@ def forward(
         else:  # To handle these `output_attentions` or `output_hidden_states` cases returning tuples
-            assert type(sa_output) == tuple
+            if type(sa_output) != tuple:
+                raise TypeError("sa_output must be a tuple but it is " + str(type(sa_output)) + " type")
```
```suggestion
                raise TypeError(f"sa_output must be a tuple but it is {type(sa_output)} type")
```
same as above!

younesbelkada commented on src/transformers/models/distilbert/modeling_distilbert.py:
```diff
@@ -348,11 +360,18 @@ def forward(
-                assert len(layer_outputs) == 2
+                if len(layer_outputs) != 2:
+                    raise ValueError(
+                        "The length of the layer_outputs should be 2, but it is " + str(len(layer_outputs))
```
```suggestion
                       f"The length of the layer_outputs should be 2, but it is {len(layer_outputs)}"
```

younesbelkada commented on src/transformers/models/distilbert/modeling_distilbert.py:
```diff
@@ -348,11 +360,18 @@ def forward(
-                assert len(layer_outputs) == 1
+                if len(layer_outputs) != 1:
+                    raise ValueError(
+                        "The length of the layer_outputs should be 1, but it is " + str(len(layer_outputs))
```
```suggestion
                        "The length of the layer_outputs should be 1, but it is {len(layer_outputs)}"
```

younesbelkada commented on src/transformers/models/distilbert/modeling_distilbert.py:
```diff
@@ -810,7 +829,9 @@ def __init__(self, config: PretrainedConfig):
         self.qa_outputs = nn.Linear(config.dim, config.num_labels)
-        assert config.num_labels == 2
+        if config.num_labels != 2:
+            raise ValueError("config.num_labels should be 2, but it is " + str(config.num_labels))
```
```suggestion
            raise ValueError("config.num_labels should be 2, but it is {config.num_labels)}"
```

JuheonChu replied:
I think you missed the opening parenthesis, but I sincerely appreciate this suggestion!

younesbelkada replied:
Ah yes you are right I missed the parenthesis here ! :)

younesbelkada reviewed (commented):
Thanks a lot for working on this! 💪 
Left a couple of comments ;) we prefer to use f-strings instead, you can check for instance [what is done in OPT](https://github.com/huggingface/transformers/blob/main/src/transformers/models/opt/modeling_opt.py#L139-L141)

JuheonChu commented:
Thank you @younesbelkada  for providing us with valuable suggestions! We will make changes and make anotehr PR! 
Sincerely appreciate it :)

younesbelkada commented:
No worries @JuheonChu ! 
As I made you some suggestions, actually you can directly continue on this PR, let me guide you through this step by step: 
Step 1: Go to "file changed" (top right of the github UI of this PR):
<img width="921" alt="Screenshot 2022-11-25 at 22 17 47" src="https://user-images.githubusercontent.com/49240599/204059724-181f2dbc-9e09-4031-bb8c-3499b3ab12ce.png">

Step 2: For each suggestion, click on "Add suggestion to batch", to add each suggestion
<img width="921" alt="Screenshot 2022-11-25 at 22 18 59" src="https://user-images.githubusercontent.com/49240599/204059801-2e1f1a1a-cf9b-43a9-8e9b-bf8d43ea4b84.png">

Step 3: Once you have added all the suggestions (make sure to add all of them), a pop up bar will appear on the top-right corner, and you can just click to it, and the suggestions will be pushed ;) 

<img width="1467" alt="Screenshot 2022-11-25 at 22 19 43" src="https://user-images.githubusercontent.com/49240599/204059837-7f8bf1fc-4c63-4234-8dae-48b52b38eb20.png">

This way no need to open a new PR each time we do a suggestion ;) ! Let me know if anything is unclear!

JuheonChu committed:
Apply suggestions from code review line 148

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

JuheonChu committed:
Apply suggestions from code review line 266

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

JuheonChu committed:
Apply suggestions from code review line 304

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

JuheonChu committed:
Apply suggestions from code review (line 365)

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

JuheonChu committed:
Apply suggestions from code review (line 373)

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

JuheonChu committed:
Update src/transformers/models/distilbert/modeling_distilbert.py

raise ValueError ( f"config.num_labels should be 2, but it is {(config.num_labels)}")

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

JuheonChu commented on src/transformers/models/distilbert/modeling_distilbert.py:
```diff
@@ -810,7 +829,9 @@ def __init__(self, config: PretrainedConfig):
         self.qa_outputs = nn.Linear(config.dim, config.num_labels)
-        assert config.num_labels == 2
+        if config.num_labels != 2:
+            raise ValueError(f"config.num_labels should be 2, but it is {config.num_labels)}"
```
```suggestion
            raise ValueError ( f"config.num_labels should be 2, but it is {(config.num_labels)}")

JuheonChu reviewed (commented):
If we change Line 833 with raise ValueError ( f"config.num_labels should be 2, but it is {(config.num_labels)}"), I think the checks will all pass!

JuheonChu committed:
Update src/transformers/models/distilbert/modeling_distilbert.py

JuheonChu commented:
Due to the reformatting of jupyter files, I created a new PR [here](https://github.com/huggingface/transformers/pull/20463).
Now it passes all the validation checks. I apologize for any inconvenience, but do you mind if you can check  [here](https://github.com/huggingface/transformers/pull/20463)?

To: @younesbelkada

younesbelkada commented:
Sure! I propose to close this PR as we moved everything on the other PR ;)

JuheonChu commented:
Closing this PR. This PR is solved in #20463.
